### PR TITLE
A version of `findIndex` for the last matching element.

### PR DESCRIPTION
**Pull request by leftaroundabout**
Discussed in

https://github.com/haskell/vector/issues/172

### CODE REVIEW

File: Data/Vector/Generic.hs [hunk 1]
```diff
@@ -1455,6 +1455,12 @@ findIndex :: Vector v a => (a -> Bool) -> v a -> Maybe Int
 {-# INLINE findIndex #-}
 findIndex f = Bundle.findIndex f . stream
 
+-- | /O(n)/ Yield 'Just' the index of the /last/ element matching the predicate
+-- or 'Nothing' if no such element exists.
+findIndexR :: Vector v a => (a -> Bool) -> v a -> Maybe Int
+{-# INLINE findIndexR #-}
```
Comment (leftaroundabout) [inline Data/Vector/Generic.hs]: I'm actually not sure if it makes sense to inline this. [dolio remarked](https://github.com/haskell/vector/issues/172) that this function won't be able to use any stream fusion. I know too little about vector's optimisations to judge this.

**Comment by cartazio** [comment]
1) could you also add a benchmark comparing this vs explicitly looping over the vector too? (we've established mostly that this isn't going to fuse with the current design, not how to make it most performant)

2) should the name be `findLastIndex` rather than `findIndexR` ?  
It seems to me that because streams and vector indices are "oriented", 'findLastIndex' might be better :)

**Comment by leftaroundabout** [comment]
1. Yeah, I can add such a benchmark. Do you mean, I should add two new function checks to `benchmarks/Main.hs`' `defaultMain` list?

2. I don't have much of an opinion on the naming. `findLastIndex` is certainly nice descriptive. But `findIndexR` should IMO be clear enough too. And it's consistent with [`Data.Sequence.findIndexR`](http://hackage.haskell.org/package/containers-0.5.10.2/docs/Data-Sequence.html#v:findIndexR) as well `streamR` on which it is based.

**Comment by leftaroundabout** [comment]
The benchmark suggests that this function is indeed worthwhile, in that it's 40% faster than a manual loop (it's possible that I've missed some optimisation that should be used in the loop, though. But then, that could also happen to other people, so...)
```
benchmarking findIndexR
time                 592.0 μs   (589.8 μs .. 594.7 μs)
                     1.000 R²   (1.000 R² .. 1.000 R²)
mean                 591.3 μs   (589.8 μs .. 593.0 μs)
std dev              5.563 μs   (4.182 μs .. 7.344 μs)

benchmarking findIndexR_naïve
time                 15.87 ms   (15.66 ms .. 16.08 ms)
                     0.999 R²   (0.998 R² .. 1.000 R²)
mean                 16.01 ms   (15.92 ms .. 16.17 ms)
std dev              316.9 μs   (207.2 μs .. 490.3 μs)

benchmarking findIndexR_manual
time                 977.0 μs   (968.0 μs .. 987.1 μs)
                     0.999 R²   (0.999 R² .. 1.000 R²)
mean                 978.0 μs   (973.1 μs .. 986.6 μs)
std dev              21.65 μs   (13.65 μs .. 34.29 μs)
```

File: benchmarks/Algo/FindIndexR.hs [hunk 2]
```diff
@@ -0,0 +1,24 @@
...
+where
+
+import Data.Vector.Unboxed (Vector)
+import qualified Data.Vector.Generic as V
+
+findIndexR :: (Double -> Bool, Vector Double) -> Maybe Int
+{-# NOINLINE findIndexR #-}
+findIndexR = uncurry V.findIndexR
+
+findIndexR_naïve :: (Double -> Bool, Vector Double) -> Maybe Int
```
Comment (Bodigrim) [inline benchmarks/Algo/FindIndexR.hs:11]: Let's avoid Unicode in function names.

**Comment by leftaroundabout** [comment]
@Bodigrim done

**Comment by lehins** [comment]
This looks good to me. @Bodigrim do you have any more suggestions? The ones you've pointed out have already been taken care of. I'll merge the PR in a few days if I don't hear form you.

**Pull request merged**